src/FSharp.Stats/ML/Unsupervised/KNN.fs

```diff
@@ -0,0 +1,170 @@
+                None
+            else            
+                let distanceIndices=
+                    points
+                    |> Seq.mapi (fun idx p -> idx, distance p x)
+                
+                let kNearestNeighborIndices =
+                    distanceIndices
+                    |> Seq.sortBy snd // snd = distance value
+                    |> Seq.take k
+                        
+                let label =
+                    kNearestNeighborIndices
+                    |> Seq.countBy (fun (idx, _) -> Seq.item idx labels)
+                    |> Seq.maxBy fst
+                    |> fst
+
+                Some label
+
+
+    /// <summary>
+    /// The [k-nearest neighbors algorithm](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm) to classify new data points into their target classes, 
+    /// depending on the features of their neighboring data points.
+    /// </summary>
+    /// <remarks>Convencience methods for using `KNN.Array.predict`, similiar to `KNeighborsClassifier` in `sklearn.neighbors`.</remarks>
+    /// <param name="distance">the distance function, e.g. `euclidean`</param>
+    /// <param name="k">The number of nearest neighbors from x to look for.</param>
+    /// <returns>The most common label from the k nearest neighbors for x.</returns>
+    /// <example> 
+    /// <code>
+    /// let reds  = [| [ 2.0; 4.0 ]; [ 1.0; 3.0 ]; [ 2.0; 4.0 ]; [ 3.0; 2.0 ]; [ 2.0; 1.0 ] |]
+    /// let blues = [| [ 5.0; 6.0 ]; [ 4.0; 5.0 ]; [ 4.0; 6.0 ]; [ 6.0; 6.0 ]; [ 5.0; 4.0 ] |]
+    /// 
+    /// let knnClassifier = KNN.Classifier(FSharp.Stats.DistanceMetrics.euclidean, 3)
+    /// 
+    /// // fit the classifier and predict new points
+    /// // version 1.
+    /// let labeledPoints = Map [ "blue", blues; "red", reds ]
+    /// knnClassifier.fit(labeledPoints)
+    /// let color  = knnClassifier.predict [3.0; 3.0] // should be: Some "red"
+    /// let colors = knnClassifier.predict [| [3.0; 3.0]; [6.0; 6.0] |] // should be: [| Some "red", Some "blue" |]
+    /// 
+    /// // version 2.
+    /// let points = Array.append reds blues
+    /// let labels = [| "red"; "red"; "red"; "red"; "red"; "blue"; "blue"; "blue"; "blue"; "blue" |]
+    /// knnClassifier.fit(points, labels)
+    /// let color  = knnClassifier.predict [3.0; 3.0] // should be: Some "red"
+    /// let colors = knnClassifier.predict [| [3.0; 3.0]; [6.0; 6.0] |] // should be: [| Some "red", Some "blue" |]
+    /// </code> 
+    /// </example>
+    type Classifier<'a, 'l when 'l: equality and 'l: comparison>(distance: Distance<'a>, k: int) =
+
+        [<DefaultValue>] val mutable labeledPoints : LabeledPoint<'a, 'l> array
+        member val K = k with get, set
+
+        member this.OverwriteK k =
+            this.K <- k
+
+        member this.fit(lps : LabeledPoint<'a, 'l> array) =
+            this.labeledPoints <- lps
+
+        member this.fit(points : 'a array, labels : 'l array) =
```

Comment:
Suggesting adding a comment that explains it will fail if both arrays aren't of same length

Reply:
Good point, will add it

Reply:
Added XML comments